### PR TITLE
Be better at recognising a terminal

### LINKED ISSUE #477: pack build ... | more produces garbage output
When pulling an image, `pack` will show a `docker` progress bar if the output goes to a terminal.

The detection of terminals could be improved.

One way to trigger an issue that is to run `pack build test --builder heroku/buildpacks | more`. It'll produce something like:

```
latest: Pulling from heroku/buildpacks

^M2746a4a261c9: Already exists ^M
^M4c1d20cdee96: Already exists ^M
^M0d3160e1d0de: Already exists ^M
^Mc8e37668deea: Already exists ^M
^Mf34fd80ed36a: Already exists ^M
^M7557d8df7b86: Already exists ^M
...
^M4f4fb700ef54: Pulling fs layer ^M^Mf88801fb2913: Waiting ^M^M72341b8babd8: Waiting ^M^Me5d4644f9d73: Waiting ^M^Mebff0555114d: Waiting ^M^M564159af21d3: Waiting ^M^Mbe24bcd9fa24: Waiting ^M^Mc070af47ae19: Waiting ^M^M1ed9dafce5c0: Waiting ^M^M2d49022bb300: Waiting ^M^Ma50d9cf1d75f: Waiting ^M^M8b85a0ecdf5b: Waiting ^M^M443d9970408b: Waiting ^M^Mdb66d6c53714: Waiting ^M^Mcc3082767d60: Waiting ^M^M7c0b72477723: Waiting ^M^M4f4fb700ef54: Waiting ^M^Ma34b15b58860: Waiting ^M^M1a98db2a454a: Downloading [=====>                                             ]     473B/4.61kB^M^M1a98db2a454a: Verifying Checksum ^M^M1a98db2a454a: Download complete ^M^M7bf5bade22e9: Downloading [>                                                  ]  539.3kB/229.9MB^M^M0e9b963bdc89: Downloading [=====================================>             ]     474B/625B^M^M0e9b963bdc89: Downloading [==================================================>]     625B/625B
^M^M0e9b963bdc89: Verifying Checksum ^M^M0e9b963bdc89: Extracting [==================================================>]     625B/625B^M^M0e9b963bdc89: Extracting [==================================================>]     625B/625B^M^M7bf5bade22e9: Downloading [>
```

### PR DESCRIPTION
**Pull request by dgageot**
Fixes #477

Signed-off-by: David Gageot <david@gageot.net>

### CODE REVIEW

**Comment by zmackie** [comment]
@dgageot Thanks for the PR!

I just ran your example from the isse in a terminal, and i'm still seeing escape codes in my terminal. Am I doing something wrong here in validating?

```bash
  |2.6.3| NY-Floater-15565 in ~/workspace/pack
± zm |fix-477 ✓| → ./out/pack build gcr.io/cncf-buildpacks-ci/pm/issues/pack:478-asdas --publish -p ../samples/apps/java-maven/ | morebionic: Pulling from cloudfoundry/cnb
dbe632be70ad: ESC[90mPulling fs layerESC[0m
750afeac99ff: ESC[90mPulling fs layerESC[0m
59bb86436456: ESC[90mPulling fs layerESC[0m
71cd62cf076a: ESC[90mPulling fs layerESC[0m
1473b2bc3374: ESC[90mPulling fs layerESC[0m
5d932446918a: ESC[90mPulling fs layerESC[0m
34e057da544e: ESC[90mPulling fs layerESC[0m
81d29541bb33: ESC[90mPulling fs layerESC[0m
ea92060a149b: ESC[90mPulling fs layerESC[0m
2b972f2ebc14: ESC[90mPulling fs layerESC[0m
60d636e99ae7: ESC[90mPulling fs layerESC[0m
e0bce984d18b: ESC[90mPulling fs layerESC[0m
859c15d89350: ESC[90mPulling fs layerESC[0m
1de1a037cfdb: ESC[90mPulling fs layerESC[0m
6f5f86374a9a: ESC[90mPulling fs layerESC[0m
7155233ef50e: ESC[90mPulling fs layerESC[0m
fc37f1c27624: ESC[90mPulling fs layerESC[0m
76cfdf8abddb: ESC[90mPulling fs layerESC[0m
0c0ed785896b: ESC[90mPulling fs layerESC[0m
8800d7e6bf7f: ESC[90mPulling fs layerESC[0m
fe65cafcdde2: ESC[90mPulling fs layerESC[0m
9648533dd025: ESC[90mPulling fs layerESC[0m
506d235b79b8: ESC[90mPulling fs layerESC[0m
64b5d8d5b04a: ESC[90mPulling fs layerESC[0m
71f8ed378449: ESC[90mPulling fs layerESC[0m
2c7f9a1dc9fa: ESC[90mPulling fs layerESC[0m
321334ddfb18: ESC[90mPulling fs layerESC[0m
8519192f8be1: ESC[90mPulling fs layerESC[0m
e6868132b93e: ESC[90mPulling fs layerESC[0m
053a4075cf39: ESC[90mPulling fs layerESC[0m
46f1f91c8481: ESC[90mPulling fs layerESC[0m
35bd12e62d75: ESC[90mPulling fs layerESC[0m
17b9144ec30d: ESC[90mPulling fs layerESC[0m
884003e65af1: ESC[90mPulling fs layerESC[0m
```

**Comment by dgageot** [comment]
@zmackie Strange... I've been testing and it's fine for me. I'm on a Mac and you?

**Comment by zmackie** [comment]
I'm also on a mac.. 🤔 
```bash
± zm |fix-477 ✓| → bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin18)
Copyright (C) 2007 Free Software Foundation, Inc.
```

**Comment by dgageot** [comment]
@zmackie I think I understand. I've got an alias to `less` for `more`. That explains the difference. With a standard `more`, you should still see the color codes being used but the progress bars should be disabled by this PR.

File: internal/image/fetcher.go [hunk 1]
```diff
@@ -7,15 +7,17 @@ import (
...
-	"github.com/buildpacks/imgutil/remote"
-	"github.com/buildpacks/lifecycle/auth"
 	"github.com/docker/docker/api/types"
 	"github.com/docker/docker/client"
 	"github.com/docker/docker/pkg/jsonmessage"
 	"github.com/google/go-containerregistry/pkg/authn"
 	"github.com/pkg/errors"
+	"golang.org/x/crypto/ssh/terminal"
+
+	"github.com/buildpacks/imgutil"
```
Comment (ekcasey) [inline internal/image/fetcher.go]: nit: the convention in this repo is to group imports from `buildpacks/*` with the external imports unless they are from `/buildpacks/pack/*`

Comment (dgageot) [reply internal/image/fetcher.go]: Thanks! Force pushed the formatting

**Comment by zmackie** [comment]
@dgageot  this looks great with the `--no-color` flag

```bash
± zm |fix-477 ✓| → ./out/pack --no-color build gcr.io/cncf-buildpacks-ci/pm/issues/pack:478-asdas --publish -p ../samples/apps/java-maven/ | more
bionic: Pulling from cloudfoundry/cnb
dbe632be70ad: Pulling fs layer
750afeac99ff: Pulling fs layer
59bb86436456: Pulling fs layer
71cd62cf076a: Pulling fs layer
1473b2bc3374: Pulling fs layer
5d932446918a: Pulling fs layer
34e057da544e: Pulling fs layer
81d29541bb33: Pulling fs layer
ea92060a149b: Pulling fs layer
c995ce118c19: Pulling fs layer
17c50dd57945: Pulling fs layer
1de1a037cfdb: Pulling fs layer
68d2c99148bc: Pulling fs layer
e0bce984d18b: Pulling fs layer
c09e86e44715: Pulling fs layer
696fd8bb65f4: Pulling fs layer
506d235b79b8: Pulling fs layer
c052d13e6e8e: Pulling fs layer
af35de5ee159: Pulling fs layer
```

**Comment by ekcasey** [comment]
I think the escape codes @zmackie was seeing are color codes, this PR looks good when `pack` is run with `--no-color`. As a separate issue we should make `pack` default to no-color when the output is not a terminal.

**Pull request merged**